### PR TITLE
feat(cmf): route onEnter/onLeave with dispatch

### PR DESCRIPTION
**Pull request by jsomsanith**
**What is the problem this PR is trying to solve?**
This is not possible to dispatch an action in cmf router onEnter/onLeave hook.

**What is the chosen solution to this problem?**
Reshape the onEnter/onLeave functions to pass redux dispatch function

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[x] This PR introduces a breaking change**
The onEnter/onLeave functions arguments have changed

### CODE REVIEW

File: output/cmf.eslint.txt [hunk 1]
```diff
@@ -22,7 +22,10 @@ The react/require-extension rule is deprecated. Please use the import/extensions
   177:51  error    There should be no spaces inside this paren  space-in-parens
 
 /home/travis/build/Talend/ui/packages/cmf/src/sagas/collection.js
-  9:1  error  Prefer default export  import/prefer-default-export
+  10:1  error  Prefer default export  import/prefer-default-export
 
-✖ 8 problems (3 errors, 5 warnings)
+/home/travis/build/Talend/ui/packages/cmf/src/UIRouter.js
+  27:78  error  'dispatch' is missing in props validation  react/prop-types
```
Comment (jare-talend) [inline output/cmf.eslint.txt]: can you clean the lint errors ?

File: output/containers.eslint.txt [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
+> @talend/react-containers@0.156.0 lint:es /home/travis/build/Talend/ui/packages/containers
 > eslint --config ../../.eslintrc src
 
 The react/require-extension rule is deprecated. Please use the import/extensions rule from eslint-plugin-import instead.
 
 /home/travis/build/Talend/ui/packages/containers/src/ActionSplitDropdown/ActionSplitDropdown.test.js
   4:10  error  'ActionSplitDropdown' is defined but never used  no-unused-vars
 
+/home/travis/build/Talend/ui/packages/containers/src/DeleteResource/deleteResource.sagas.js
+  106:4  warning  Unexpected console statement  no-console
```
Comment (jare-talend) [inline output/containers.eslint.txt]: can you remove the console

File: output/cmf.eslint.txt (hunk 1 above)
Comment (jsomsanith) [reply output/cmf.eslint.txt]: Oh, I haven't turned on eslint on my new computer, sorry for that

File: output/containers.eslint.txt (hunk 2 above)
Comment (jsomsanith) [reply output/containers.eslint.txt]: WTF Oo, haven't touched this file
I'll fix it

File: packages/cmf/src/UIRouter.js [hunk 3]
```diff
@@ -25,23 +24,20 @@ import api from './api';
  * @return {object} ReactElement
  */
 const UIRouter = (props, context) => {
-	const routes = api.route.getRoutesFromSettings(context, props.routes);
-	if (routes.path === '/' && !!routes.component) {
-		return (<BaseRouter routes={routes} history={props.history} />);
+	const routes = api.route.getRoutesFromSettings(context, props.routes, props.dispatch);
```
Comment (jmfrancois) [inline packages/cmf/src/UIRouter.js]: The UIRouter is just connected, not cmfConnected so you don't have props.dispatch.
You may have it from the context  if you add the store to it

Comment (jsomsanith) [reply packages/cmf/src/UIRouter.js]: Yes but it's ok IMO. The entry point to cmf store is still redux, so dispatching an action from react-redux is the same as dispatching an action from cmf redux.

File: packages/cmf/src/settings.md [hunk 4]
```diff
@@ -130,3 +130,49 @@ CMF uses [React router](https://github.com/ReactTraining/react-router). The defi
...
+const registerRouteFunction = api.route.registerFunction;
+registerRouteFunction('my:route:onEnter', onEnterFetchThings);
+```
+
+```json
+// settings.json
+{
+  "routes": {
+    "path": "/",
+    ...
```
Comment (frassinier) [inline packages/cmf/src/settings.md]: stop using three dots, looks weird

File: packages/cmf/src/settings.md [hunk 5]
```diff
@@ -130,3 +130,49 @@ CMF uses [React router](https://github.com/ReactTraining/react-router). The defi
   }
 }
 ```
+
+You can use onEnter/onLeave lifecycle hooks. To do that, you need to
```
Comment (frassinier) [inline packages/cmf/src/settings.md]: You can use onEnter/onLeave lifecycle hooks. To do that, you need to:

File: packages/cmf/src/settings.md [hunk 6]
```diff
@@ -130,3 +130,49 @@ CMF uses [React router](https://github.com/ReactTraining/react-router). The defi
   }
 }
 ```
+
+You can use onEnter/onLeave lifecycle hooks. To do that, you need to 
+* register a route function in CMF registry
+* add the route function id in your route setting
```
Comment (frassinier) [inline packages/cmf/src/settings.md]: dd the route function id in your route settings

**Pull request merged**